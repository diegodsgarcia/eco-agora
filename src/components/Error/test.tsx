import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helper'

import Error from '.'

test('Should render the component', () => {
  const message = 'Ops! Ocorreu algum problema'
  const { container } = renderWithTheme(<Error message={message} />)

  expect(screen.getByText(/ops! ocorreu algum :\(/i)).toBeInTheDocument()
  expect(screen.getByText(message)).toBeInTheDocument()

  expect(container.firstChild).toMatchSnapshot()
})
