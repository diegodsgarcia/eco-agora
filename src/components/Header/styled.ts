import styled, { css } from 'styled-components'
export const Wrapper = styled.header`
  display: flex;
  height: 5rem;
`

export const Logo = styled.h1`
  ${({ theme }) => css`
    display: flex;
    font-size: 2.5rem;
    flex: 1;
    align-items: center;
    justify-content: center;
    color: ${theme.colors.primary};
  `}
`

type MenuProps = {
  open: boolean
}

export const Menu = styled.div<MenuProps>`
  ${({ open }) => css`
    position: absolute;
    right: 0;
    display: flex;
    gap: 0.8rem;
    flex-direction: column;
    width: 5rem;
    height: 5rem;
    padding-right: 1rem;
    padding-top: 1rem;
    cursor: pointer;
    z-index: 3;

    ${open &&
    css`
      ${Line}:nth-child(1) {
        transform: translateY(13px) rotate(45deg);
        transform-origin: center;
      }

      ${Line}:nth-child(2) {
        opacity: 0;
      }

      ${Line}:nth-child(3) {
        transform: translateY(-19px) rotate(-45deg);
        transform-origin: center;
      }
    `}
  `}
`

export const Line = styled.div`
  ${({ theme }) => css`
    height: 8px;
    background: ${theme.colors.white};
    opacity: 1;
    transition: 0.3s ease-out;

    &:nth-child(2) {
      transition: 0.1s ease-out;
    }
  `}
`

export const Nav = styled.nav<MenuProps>`
  ${({ theme, open }) => css`
    position: absolute;
    display: flex;
    flex-direction: column;
    padding-top: 5rem;
    opacity: 1;
    visibility: visible;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: ${theme.colors.background};
    transition: 0.3s ease-out;
    z-index: 2;

    ${!open &&
    css`
      opacity: 0;
      visibility: hidden;
    `}
  `}
`

type NavItemProps = {
  isSelected: boolean
}

export const NavItem = styled.a<NavItemProps>`
  ${({ theme, isSelected }) => css`
    padding: 1rem;
    font-size: 1.8rem;
    cursor: pointer;
    color: ${isSelected ? theme.colors.primary : theme.colors.white};

    &:hover {
      color: ${theme.colors.primary};
      background-color: ${theme.colors.backgroundHover};
    }
  `}
`
