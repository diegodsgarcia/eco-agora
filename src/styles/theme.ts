export default {
  colors: {
    primary: '#ad4',
    background: '#232323',
    backgroundHover: '#1b1b1b',
    white: '#ffffff',
    black: '#000000'
  }
} as const
