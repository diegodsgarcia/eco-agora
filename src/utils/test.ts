import { LocaleAPI } from 'types/Locale'
import { buildLocale, convertDate } from 'utils'

test('Should build Locale', () => {
  const mock: LocaleAPI = {
    Cameras: [
      {
        DisplayName: 'Anchieta',
        ExternalCode: 'km 102',
        Cameras: [
          {
            CameraID: 1,
            KM: 12,
            Description: 'km 102',
            LastUpdate: 'Date(1212334324)',
            Url: ''
          }
        ]
      }
    ]
  }

  expect(buildLocale(mock)).toStrictEqual([
    {
      name: mock.Cameras[0].DisplayName,
      externalCode: mock.Cameras[0].ExternalCode,
      cameras: [
        {
          id: mock.Cameras[0].Cameras[0].CameraID,
          km: mock.Cameras[0].Cameras[0].KM,
          description: mock.Cameras[0].Cameras[0].Description,
          lastUpdate: mock.Cameras[0].Cameras[0].LastUpdate,
          url: mock.Cameras[0].Cameras[0].Url
        }
      ]
    }
  ])
})

test('Should build void Locale', () => {
  const mock: LocaleAPI = {
    Cameras: []
  }

  expect(buildLocale(mock)).toStrictEqual([])
})

test('Should be convert to Date', () => {
  const mock = 'Date(1609259906000)'

  expect(convertDate(mock)).toStrictEqual('29/12/2020 13:38:26')
})
