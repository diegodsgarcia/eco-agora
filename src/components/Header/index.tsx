import { useState } from 'react'

import { Locale } from 'types/Locale'

import * as S from './styled'

export type HeaderProps = {
  menuItems: Locale[]
  itemSelected?: Locale
  onMenuClick: (locale: Locale) => void
}

const Header = ({ menuItems = [], itemSelected, onMenuClick }: HeaderProps) => {
  const [menuIsOpen, setMenuIsOpen] = useState(false)
  return (
    <S.Wrapper>
      <S.Logo>Eco Agora</S.Logo>
      {!!menuItems.length && (
        <S.Menu
          aria-label="Menu button"
          role="button"
          open={menuIsOpen}
          onClick={() => {
            setMenuIsOpen(!menuIsOpen)
          }}
        >
          <S.Line />
          <S.Line />
          <S.Line />
        </S.Menu>
      )}

      <S.Nav open={menuIsOpen} aria-hidden={!menuIsOpen}>
        {menuItems.map((item, i) => (
          <S.NavItem
            key={`nav-${i}`}
            isSelected={item === itemSelected}
            onClick={() => {
              setMenuIsOpen(false)
              onMenuClick(item)
            }}
          >
            {item.name}
          </S.NavItem>
        ))}
      </S.Nav>
    </S.Wrapper>
  )
}

export default Header
