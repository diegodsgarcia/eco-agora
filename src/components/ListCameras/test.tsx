import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helper'
import { css } from 'styled-components'

import ListCameras from '.'
import * as S from './styled'
import camerasMock from './mock'

describe('<ListCameras />', () => {
  it('should render list of cameras', async () => {
    renderWithTheme(<ListCameras cameras={camerasMock} />)

    expect(screen.getAllByRole('img')).toHaveLength(6)

    expect(screen.queryAllByRole('figure')[0]).toHaveStyleRule('opacity', '1', {
      modifier: `:hover ${css`
        ${S.Description}
      `}`
    })
  })
})
