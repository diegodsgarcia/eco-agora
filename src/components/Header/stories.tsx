import { Story, Meta } from '@storybook/react/types-6-0'

import Header, { HeaderProps } from '.'

export default {
  title: 'Header',
  args: {
    menuItems: [
      {
        name: 'Endereço 1',
        url: '#'
      },
      {
        name: 'Endereço 2',
        url: '#'
      },
      {
        name: 'Endereço 3',
        url: '#'
      }
    ]
  },
  parameters: {
    layout: 'fullscreen'
  }
} as Meta

export const Default: Story<HeaderProps> = (args) => <Header {...args} />
