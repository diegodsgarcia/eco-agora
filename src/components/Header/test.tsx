import { fireEvent, screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helper'

import Header, { HeaderProps } from '.'
import theme from 'styles/theme'

const props: HeaderProps = {
  menuItems: [
    {
      name: 'Endereço 1',
      externalCode: 'end-1',
      cameras: []
    },
    {
      name: 'Endereço 2',
      externalCode: 'end-2',
      cameras: []
    },
    {
      name: 'Endereço 3',
      externalCode: 'end-3',
      cameras: []
    }
  ],
  onMenuClick: jest.fn()
}

describe('<Header />', () => {
  it('should render header', () => {
    const { container } = renderWithTheme(<Header {...props} />)

    expect(
      screen.getByRole('heading', { name: /eco agora/i })
    ).toBeInTheDocument()

    expect(screen.getByLabelText(/menu button/i)).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should toggle navigation when click in menu button', () => {
    renderWithTheme(<Header {...props} />)

    const nav = screen.getByRole('navigation', { hidden: true })
    const menu = screen.getByLabelText(/menu button/i)

    expect(nav.getAttribute('aria-hidden')).toBe('true')
    expect(nav).toHaveStyle({ opacity: 0 })

    fireEvent.click(menu)

    expect(nav.getAttribute('aria-hidden')).toBe('false')
    expect(nav).toHaveStyle({ opacity: 1 })

    fireEvent.click(screen.getByText(/endereço 1/i))

    expect(props.onMenuClick).toHaveBeenCalledTimes(1)
  })

  it('should render a selected menu item', () => {
    renderWithTheme(<Header {...props} itemSelected={props.menuItems[0]} />)

    expect(screen.getByText(/endereço 1/i)).toHaveStyle({
      color: theme.colors.primary
    })
  })
})
