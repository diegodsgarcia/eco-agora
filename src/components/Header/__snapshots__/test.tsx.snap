// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render header 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 5rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 2.5rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #ad4;
}

.c2 {
  position: absolute;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.8rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 5rem;
  height: 5rem;
  padding-right: 1rem;
  padding-top: 1rem;
  cursor: pointer;
  z-index: 3;
}

.c3 {
  height: 8px;
  background: #ffffff;
  opacity: 1;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
}

.c3:nth-child(2) {
  -webkit-transition: 0.1s ease-out;
  transition: 0.1s ease-out;
}

.c4 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-top: 5rem;
  opacity: 1;
  visibility: visible;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: #232323;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
  z-index: 2;
  opacity: 0;
  visibility: hidden;
}

.c5 {
  padding: 1rem;
  font-size: 1.8rem;
  cursor: pointer;
  color: #ffffff;
}

.c5:hover {
  color: #ad4;
  background-color: #1b1b1b;
}

<header
  class="c0"
>
  <h1
    class="c1"
  >
    Eco Agora
  </h1>
  <div
    aria-label="Menu button"
    class="c2"
    role="button"
  >
    <div
      class="c3"
    />
    <div
      class="c3"
    />
    <div
      class="c3"
    />
  </div>
  <nav
    aria-hidden="true"
    class="c4"
  >
    <a
      class="c5"
    >
      Endereço 1
    </a>
    <a
      class="c5"
    >
      Endereço 2
    </a>
    <a
      class="c5"
    >
      Endereço 3
    </a>
  </nav>
</header>
`
